mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala

```diff
@@ -311,22 +346,39 @@ class VectorIndexerModel private[ml] (
   // TODO: Check more carefully about whether this whole class will be included in a closure.
 
   /** Per-vector transform function */
-  private val transformFunc: Vector => Vector = {
+  private lazy val transformFunc: Vector => Vector = {
     val sortedCatFeatureIndices = categoryMaps.keys.toArray.sorted
     val localVectorMap = categoryMaps
     val localNumFeatures = numFeatures
+    val localHandleInvalid = getHandleInvalid
     val f: Vector => Vector = { (v: Vector) =>
       assert(v.size == localNumFeatures, "VectorIndexerModel expected vector of length" +
         s" $numFeatures but found length ${v.size}")
       v match {
         case dv: DenseVector =>
+          var hasInvalid = false
           val tmpv = dv.copy
           localVectorMap.foreach { case (featureIndex: Int, categoryMap: Map[Double, Int]) =>
-            tmpv.values(featureIndex) = categoryMap(tmpv(featureIndex))
+            try {
```

Comment:
Is it really faster to use try-catch?  I've always understood it to be better to avoid Exceptions for implementing logic; e.g., here, we could check ```categoryMap.contains(tmpv(featureIndex))``` or use ```getOrElse```.

Reply:
Ah, that's better. Guess Xiangrui's comments become our preoccupation. Thanks.

Reply:
But, I don't think so. JVM exception handling will be very efficient, when exception do not occur, the code with `try... catch` will have the same performance with non `try... catch` code.
here is some explanation: https://www.quora.com/How-expensive-is-the-try-catch-block-in-Java-in-terms-of-performance

Reply:
In this scenario, we can assume that, invalid values rarely occur.

Reply:
The try part is fast, yet the catch part can be very slow comparably.

Reply:
Yes. If exception occur, it will be slow, but, I think we can assume the exception will rarely occur, isn't it ?